Added buttons for previous, next and reset, and keys [ for previous a…

FlorisE opened the pull request:
Added buttons for previous, next and reset (go to first training view), and keys [ for previous and ] for next

We also show the name of the image for the training view (updated whenever the view is changed to a training view)

We found these buttons/keys quite useful for stepping through datasets to find bad cameras.

Tom94 commented:
Hi there, thanks for the contribution.

There are a few cases, where `dataset.paths` doesn't get populated properly, leading to undefined behavior in the GUI. For example, when serializing / deserializing a dataset (see `json_binding.h`) or when using `create_empty_nerf_dataset`.

It would also be good if the GUI used `.at()` instead of `operator[]` to access the `paths` array to more easily catch such cases.

Cheers!

FlorisE commented:
Thanks for your comment, I now populate the list with empty strings in the `create_empty_nerf_dataset` case. I also switched to using `.at()` for getting the path from the vector of paths. For now I added an empty catch statement, is that OK?

In the case of from_json and to_json, should I also save/load the vector of paths, or at least initialize empty values for each view?

Tom94 commented:
Thanks for the update! No need for the empty catch statement -- I think it's fine if the program crashes as long as it's not undefined behavior but an easily understandable exception.

For `from_json` and `to_json`: yes, please. Storing the paths should be as simple as adding `j["paths"] = dataset.paths;`, whereas loading should initialize to `j["paths"]` if the key is present and empty strings otherwise. Cheers!

FlorisE commented:
I removed the try/catch structure in `testbed.cc` and now set the paths in `from_json` and `to_json` in `json_binding.h`. I'm not sure at what time these functions are called so I didn't test the code, do you have some advice on how to test these functions?

Tom94 commented:
Thank you!

They're getting used when (i) saving a snapshot of a trained model and (ii) when loading a snapshot of a trained model without a dataset specified. For an example of (ii):

`./run.py --gui --mode nerf --load_snapshot <snapshot filename>.msgpack`

Tom94 commented on include/neural-graphics-primitives/json_binding.h:
```diff
@@ -175,6 +176,11 @@ inline void from_json(const nlohmann::json& j, NerfDataset& dataset) {
 			from_json(ji.at("camera_distortion"), dataset.metadata[i].camera_distortion);
 		}
+		if (j.contains("paths")) {
+			dataset.paths.push_back(j["paths"].at(i));
```
I recommend not using `push_back` in case `from_json` is called multiple times on the same dataset (which would result in an incorrect number of paths). Instead, above the for loop, I would do the following:

```cpp
dataset.paths.resize(dataset.n_images, "");
if (j.contains("paths")) dataset.paths = j["paths"];
```

Tom94 replied:
that's perfect!

Tom94 commented:
There's a local variable `bool shift` that you can use in place of `io.KeyShift` -- sounds good otherwise

FlorisE commented:
Updated the code to use the local `shift` variable instead.
